src/CFamily/CMake/VsDevCmdEnvironmentProvider.cs

```diff
@@ -0,0 +1,231 @@
+{
+    internal interface IVsDevCmdEnvironmentProvider
+    {
+        /// <summary>
+        /// Returns the environment settings set by VsDevCmd.bat
+        /// </summary>
+        /// <param name="scriptParams">Any additional parameters to pass to the batch file. Can be null.</param>
+        IReadOnlyDictionary<string, string> Get(string scriptParams);
+    }
+
+    internal class VsDevCmdEnvironmentProvider : IVsDevCmdEnvironmentProvider
+    {
+        /// <summary>
+        /// Relative path from the VS installation root to the batch file
+        /// </summary>
+        /// <remarks>This path works for VS2015+</remarks>
+        private const string RelativePathToBatchFile = "Common7\\Tools\\VsDevCmd.bat";
+
+        private const int SCRIPT_TIMEOUT_MS = 4000;
+
+        private readonly IVsInfoService vsInfoService;
+        private readonly ILogger logger;
+
+        // Interfaces for testing
+        private readonly IProcessFactory processFactory;
+        private readonly IFileSystem fileSystem;
+
+
+        public VsDevCmdEnvironmentProvider(IVsInfoService vsInfoService, ILogger logger)
+            : this(vsInfoService, logger, new ProcessFactory(), new FileSystem())
+        {
+            this.vsInfoService = vsInfoService;
+            this.logger = logger;
+        }
+
+        internal /* for testing */ VsDevCmdEnvironmentProvider(IVsInfoService vsInfoService, ILogger logger,
+            IProcessFactory processFactory, IFileSystem fileSystem)
+        {
+            this.vsInfoService = vsInfoService;
+            this.logger = logger;
+            this.processFactory = processFactory;
+            this.fileSystem = fileSystem;
+        }
+
+        public IReadOnlyDictionary<string, string> Get(string scriptParams)
+        {
+            try
+            {
+                string filePath = Path.Combine(vsInfoService.InstallRootDir, RelativePathToBatchFile);
+                LogDebug($"VsDevCmd location: {filePath}");
+
+                if (!fileSystem.File.Exists(filePath))
+                {
+                    logger.WriteLine(Resources.VsDevCmd_FileNotFound, filePath);
+                    return null;
+                }
+
+                var capturedOutput = ExecuteVsDevCmd(filePath, scriptParams);
+                var settings = ParseOutput(capturedOutput);
+
+                if (settings == null || settings.Count == 0)
+                {
+                    logger.WriteLine(Resources.VsDevCmd_NoSettingsFound);
+                    return null;
+                }
+
+                LogDebug($"Settings fetched successfully. Count: {settings.Count}");
+                return settings;
+            }
+            catch (Exception ex) when (!ErrorHandler.IsCriticalException(ex))
+            {
+                logger.WriteLine(Resources.VsDevCmd_ErrorFetchingSettings, ex);
+            }
+
+            return null;
+        }
+
+        internal /* for testing */ string UniqueId { get; private set;}
```

Comment:
I'm debating if we should allow to run this more than once. The variables are cached on VS's side, so is there any benefit in supporting multiple calls? It would simplify things to fetch it in the CTOR and cache them. WDYT?

Reply:
My assumption is that we might need to call this multiple times with different 'scriptParams', so potentially we'd call this multiple times with different inputs.

We certainly won't want to run this on every call, though - it takes just under a second to run on my local machine. 
There is also the issue of some kind of synchronization mechanism to make sure we've fetched the data before the analyzers run, and this issue that the user might be able to override some of the settings in the json/cmake settings file.

However, I think all of those things need to be done outside this class. This class is just responsible for fetching the VS settings when asked.